samples/js/photos/src/photo-set-iterator.js

```diff
@@ -17,103 +18,157 @@ import type {
   NomsPhoto,
 } from './types.js';
 
-type PhotoSetIteratorResult = AsyncIteratorResult<[number /* -date */, NomsPhoto]>;
+type IterValue = [number /* -date */, NomsPhoto];
 
-export default class PhotoSetIterator extends AsyncIterator<[number, NomsPhoto]> {
+export default class PhotoSetIterator extends AsyncIterator<IterValue> {
 }
 
 export class SinglePhotoSetIterator extends PhotoSetIterator {
-  _outer: AsyncIterator<PhotoSetEntry>;
-  _start: ?NomsPhoto;
-  _inner: ?AsyncIterator<NomsPhoto>;
-  _negdate: ?number;
+  _outer: PhotoSet;
+  _startDate: number;
+  _startPhoto: ?NomsPhoto;
+  _outerIter: ?AsyncIterator<PhotoSetEntry>; // null until first call to next()
+  _outerDate: number;                        // NaN until first call to next()
+  _innerIter: ?AsyncIterator<NomsPhoto>;     // null until first call to next()
+  _done: boolean;
+  _lock: boolean;
 
   static all(photoSet: PhotoSet): SinglePhotoSetIterator {
-    return new SinglePhotoSetIterator(photoSet.iterator(), null);
+    return new SinglePhotoSetIterator(photoSet, -Infinity, null);
   }
 
-  static at(photoSet: PhotoSet, negdate: number, photo: ?NomsPhoto): SinglePhotoSetIterator {
-    return new SinglePhotoSetIterator(photoSet.iteratorAt(negdate), photo);
+  static at(photoSet: PhotoSet, startDate: number, startPhoto: ?NomsPhoto): SinglePhotoSetIterator {
+    return new SinglePhotoSetIterator(photoSet, startDate, startPhoto);
   }
 
-  constructor(outer: AsyncIterator<PhotoSetEntry>, start: ?NomsPhoto) {
+  constructor(outer: PhotoSet, startDate: number, startPhoto: ?NomsPhoto) {
     super();
     this._outer = outer;
-    this._start = start;
+    this._startDate = startDate;
+    this._startPhoto = startPhoto;
+    this._outerIter = null;
+    this._outerDate = NaN;
+    this._innerIter = null;
+    this._done = false;
+    this._lock = false;
   }
 
-  async next(): Promise<PhotoSetIteratorResult> {
-    if (!this._inner && await this._nextInner()) {
+  async next(): Promise<AsyncIteratorResult<IterValue>> {
+    invariant(!this._lock);
+    this._lock = true;
+    try {
+      return await this._next(); // must await so that finally will execute afterwards
```

Comment:
These changes are actually just a sanity check which is triggering a bug in the photo grid, triggers when you use the date slider. I'll fix that next.